quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/resteasy/TransactionalSessionHandler.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.quarkus.runtime.integration.resteasy;
+
+import io.quarkus.arc.Arc;
+
+import org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext;
+import org.jboss.resteasy.reactive.server.spi.ServerRestHandler;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.utils.KeycloakSessionUtil;
+
+public final class TransactionalSessionHandler implements ServerRestHandler, org.keycloak.quarkus.runtime.transaction.TransactionalSessionHandler {
+
+    @Override
+    public void handle(ResteasyReactiveRequestContext requestContext) {
+        requestContext.requireCDIRequestScope();
+        KeycloakSession currentSession = Arc.container().instance(KeycloakSession.class).get();
+        // this handler might be invoked multiple times when resolving sub-resources
+        // make sure the transaction is began once when the session is first associated with the thread
+        if (KeycloakSessionUtil.setKeycloakSession(currentSession) == null) {
```

Comment:
Isn't better to check `session.getTransactionManager().isActive()` instead?

One less usage of `KeycloakSessionUtil`.

Reply:
I had that before, but that's not quite the same as the previous logic - if for whatever reason the transaction was already committed or rolledback, we don't want to start another one.

Reply:
If both happen (commit or rollback) the transaction manager should return `false` from `isActive`. But it is not a blocker for moving forward if you want to re-evaluate this later if needed.

Related to https://github.com/keycloak/keycloak/issues/29107, I'm looking forward to removing this mechanism from obtaining sessions or at least removing its references as much as we can from our code base.

Reply:
> If both happen (commit or rollback) the transaction manager should return false from isActive. But it is not a blocker for moving forward if you want to re-evaluate this later if needed.

The previous logic would begin the transaction only once for a given request cycle, so if something new is needed, then yes let's revisit.

> Related to https://github.com/keycloak/keycloak/issues/29107, I'm looking forward to removing this mechanism from obtaining sessions or at least removing its references as much as we can from our code base.

Sounds good, that will help further clarify the availability of the session.